### PR TITLE
fix: increment and refactor number of hops for routed messages

### PR DESCRIPTION
**Pull request by stedfn**

### CODE REVIEW

**Review by saketh-are — APPROVED** [review]
Thanks!

Could be worth adding a couple of basic unit tests covering the 0-hop and 1-hop cases.

**Comment by stedfn** [comment]
@saketh-are I cleaned it up a bit more and added an assert in an integration test to confirm it gets incremented. Let me know if you are happy with the changes

File: chain/network/src/network_protocol/network.proto [hunk 1]
```diff
@@ -373,11 +373,14 @@ message SignedTransaction {
...
 message RoutedMessage {
+  // Deprecated
+  reserved 3;
+
   bytes borsh = 1;
   // Timestamp of creating the Routed message by its original author.
   google.protobuf.Timestamp created_at = 2;
   // Number of peers this routed message travelled through. Doesn't include the peer that created the message.
-  optional int32 num_hops = 3;
+  uint32 num_hops = 4;
```
Comment (saketh-are) [inline chain/network/src/network_protocol/network.proto:383]: Just want to confirm, are you completely sure that deprecating this field and reusing the field name is OK?

**Review by saketh-are — COMMENTED** [review]
Still LGTM, left a couple of small comments

File: chain/network/src/network_protocol/network.proto (hunk 1 above)
Comment (stedfn) [reply chain/network/src/network_protocol/network.proto:383]: It should be ok:
- reusing the field name is fine as the protobuf encoding is based on the tag numbers
- there could be some data loss with deprecating the old field (for example if a node that uses the old version sends it to one that uses the new one) but the field was not used before anyway

There also is a protobuf backwards compatibility [test](https://github.com/near/nearcore/actions/runs/11292864089/job/31409746916?pr=12188) which passed, but I do not know how it works

**Pull request merged**